File: web/app/lib/powersync-schema.ts
import { column, Schema, Table } from "@powersync/web";

const playlists = new Table(
  {
    name: column.text,
    type: column.text,
    url: column.text,
    username: column.text,
    encrypted_password: column.text,
    epg_url: column.text,
    refresh_hrs: column.integer,
    is_active: column.integer,
    sort_order: column.integer,
    last_sync: column.text,
    last_epg_sync: column.text,
    created_at: column.text,
    updated_at: column.text,
  },
  { indexes: { sort: ["sort_order"] } }
);

const channels = new Table(
  {
    playlist_id: column.text,
    source_channel_id: column.text,
    tvg_id: column.text,
    name: column.text,
    group_name: column.text,
    epg_id: column.text,
    logo_url: column.text,
    stream_url: column.text,
    channel_number: column.integer,
    is_favorite: column.integer,
    is_deleted: column.integer,
    deleted_at: column.text,
    created_at: column.text,
    updated_at: column.text,
  },
  {
    indexes: {
      playlist: ["playlist_id"],
      group: ["playlist_id", "group_name"],
      favorite: ["is_favorite"],
      name: ["name"],
    },
  }
);

const vod_items = new Table(
  {
    playlist_id: column.text,
    title: column.text,
    type: column.text,
    stream_url: column.text,
    logo_url: column.text,
    genre: column.text,
    year: column.integer,
    rating: column.text,
    duration: column.integer,
    season_num: column.integer,
    episode_num: column.integer,
    series_id: column.text,
    container_extension: column.text,
    plot: column.text,
    cast_list: column.text,
    director: column.text,
    created_at: column.text,
    updated_at: column.text,
  },
  {
    indexes: {
      playlist_type: ["playlist_id", "type"],
      title: ["title"],
      series: ["series_id"],
    },
  }
);

const watch_progress = new Table(
  {
    content_id: column.text,
    playlist_id: column.text,
    position_ms: column.integer,
    duration_ms: column.integer,
    updated_at: column.text,
  },
  { indexes: { content: ["content_id"] } }
);

const user_preferences = new Table({
  preferred_engine: column.text,
  resume_playback_enabled: column.integer,
  buffer_timeout_seconds: column.integer,
  updated_at: column.text,
});

export const AppSchema = new Schema({
  playlists,
  channels,
  vod_items,
  watch_progress,
  user_preferences,
});

export type Database = (typeof AppSchema)["types"];

File: StreamDeck/Sources/SyncDatabase/PowerSyncSchema.swift
import PowerSync

/// PowerSync schema definition matching the Supabase tables and web dashboard schema.
/// Must stay in sync with `web/app/lib/powersync-schema.ts`.
public let syncSchema = Schema(
    tables: [
        Table(
            name: "playlists",
            columns: [
                .text("name"),
                .text("type"),
                .text("url"),
                .text("username"),
                .text("encrypted_password"),
                .text("epg_url"),
                .integer("refresh_hrs"),
                .integer("is_active"),
                .integer("sort_order"),
                .text("last_sync"),
                .text("last_epg_sync"),
                .text("created_at"),
                .text("updated_at"),
            ],
            indexes: [
                Index.ascending(name: "playlists_sort", columns: ["sort_order"])
            ]
        ),
        Table(
            name: "channels",
            columns: [
                .text("playlist_id"),
                .text("source_channel_id"),
                .text("tvg_id"),
                .text("name"),
                .text("group_name"),
                .text("epg_id"),
                .text("logo_url"),
                .text("stream_url"),
                .integer("channel_number"),
                .integer("is_favorite"),
                .integer("is_deleted"),
                .text("deleted_at"),
                .text("created_at"),
                .text("updated_at"),
            ],
            indexes: [
                Index.ascending(name: "channels_playlist", columns: ["playlist_id"]),
                Index(name: "channels_group", columns: [
                    IndexedColumn.ascending("playlist_id"),
                    IndexedColumn.ascending("group_name"),
                ]),
                Index.ascending(name: "channels_favorite", column: "is_favorite"),
                Index.ascending(name: "channels_name", column: "name"),
            ]
        ),
        Table(
            name: "vod_items",
            columns: [
                .text("playlist_id"),
                .text("title"),
                .text("type"),
                .text("stream_url"),
                .text("logo_url"),
                .text("genre"),
                .integer("year"),
                .text("rating"),
                .integer("duration"),
                .integer("season_num"),
                .integer("episode_num"),
                .text("series_id"),
                .text("container_extension"),
                .text("plot"),
                .text("cast_list"),
                .text("director"),
                .text("created_at"),
                .text("updated_at"),
            ],
            indexes: [
                Index(name: "vod_playlist_type", columns: [
                    IndexedColumn.ascending("playlist_id"),
                    IndexedColumn.ascending("type"),
                ]),
                Index.ascending(name: "vod_title", column: "title"),
                Index.ascending(name: "vod_series", column: "series_id"),
            ]
        ),
        Table(
            name: "watch_progress",
            columns: [
                .text("content_id"),
                .text("playlist_id"),
                .integer("position_ms"),
                .integer("duration_ms"),
                .text("updated_at"),
            ],
            indexes: [
                Index.ascending(name: "progress_content", column: "content_id")
            ]
        ),
        Table(
            name: "user_preferences",
            columns: [
                .text("preferred_engine"),
                .integer("resume_playback_enabled"),
                .integer("buffer_timeout_seconds"),
                .text("updated_at"),
            ]
        ),
    ]
)
